Toolkit/ArcGISToolkit/AR/ArcGISARView.swift

```diff
@@ -0,0 +1,512 @@
+        }
+    }
+    
+    var errorDescription: String? {
+        switch self {
+        case .accessDenied:
+            return "Access to the device location is denied."
+        case .missingPListKey:
+            return "You must specify a location usage description key (NSLocationWhenInUseUsageDescription or NSLocationAlwaysUsageDescription) in your plist."
+        }
+    }
+}
+
+public class ArcGISARView: UIView {
+    
+    enum clError {
+        case accessDenied
+        case missingPListKey
+    }
+
+    // MARK: public properties
+    
+    /// The view used to display the `ARKit` camera image and 3D `SceneKit` content.
+    public lazy private(set) var arSCNView = ARSCNView(frame: .zero)
+    
+    /// The view used to display ArcGIS 3D content.
+    public lazy private(set) var sceneView = AGSSceneView(frame: .zero)
+    
+    /// The world tracking information used by `ARKit`.
+    public var arConfiguration: ARConfiguration = {
+        let config = ARWorldTrackingConfiguration()
+        config.worldAlignment = .gravityAndHeading
+        return config
+        }() {
+        didSet {
+            // Start tracking using the new configuration.
+            startTracking()
+        }
+    }
+
+    /// The viewpoint camera used to set the initial view of the sceneView instead of the devices GPS location via the locationManager.
+    public var originCamera: AGSCamera?
+    
+    /// The translation factor used to support a table top AR experience.
+    public var translationTransformationFactor: Double = 1.0
+    
+    // We implement ARSCNViewDelegate methods, but will use `arSCNViewDelegate` to forward them to clients.
+    weak open var arSCNViewDelegate: ARSCNViewDelegate?
+
+    // MARK: private properties
+    
+    /// Whether to display the camera image or not.
+    private var renderVideoFeed = true
+    
+    /// Used to determine the device location when originCamera is not set.
+    private lazy var locationManager: CLLocationManager = { [unowned self] in
+        let lm = CLLocationManager()
+        lm.desiredAccuracy = kCLLocationAccuracyBest
+        lm.delegate = self
+        return lm
+    }()
+    
+    /// Initial location from locationManager.
+    private var initialLocation: CLLocation?
+    
+    /// Current horizontal accuracy of the device.
+    private var horizontalAccuracy: CLLocationAccuracy = .greatestFiniteMagnitude
+    
+    /// The intial camera position and orientation whether it was set via originCamera or the locationManager.
+    private var initialTransformationMatrix = AGSTransformationMatrix()
+    
+    /// Whether `ARKit` is supported on this device.
+    private var isSupported = {
+        return ARWorldTrackingConfiguration.isSupported
+    }()
+    
+    /// Whether the client has been notfiied of start/failure.
+    private var notifiedStartOrFailure = false
+    
+    /// Used when calculating framerate.
+    private var lastUpdateTime: TimeInterval = 0
```

Comment:
for a `TimeInterval` property, I'm not sure the name `lastUpdateTIme` makes sense. Is it more "timeSinceLastUpdate`?

Reply:
`lastUpdateTime` is used in `willRenderScene`.  The `time` argument of the [willRenderScene]() method is defined as: _The current system time, in seconds. If your custom rendering involves animation, use this parameter to compute your own animation state._

`lastUpdateTime` is set to that value at the end of the method.  So it's not an interval, but an actual time.

Reply:
ok